<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Scott Barlowe" />

<meta name="date" content="2016-06-29" />

<title>SubVis</title>






<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">SubVis</h1>
<h4 class="author"><em>Scott Barlowe</em></h4>
<h4 class="date"><em>2016-06-29</em></h4>



<div id="description" class="section level3">
<h3>Description:</h3>
<pre><code>Software tool for visual analysis of substitution matrix effects on pairwise protein sequence     alignment.  Utilizes Shiny for interface components, R for alignment processing, and              JavaScript for visualization.</code></pre>
</div>
<div id="other-project-pages" class="section level3">
<h3>Other Project Pages:</h3>
<pre><code>  1.  https://github.com/sabarlowe/SubVis</code></pre>
</div>
<div id="known-issues" class="section level3">
<h3>Known Issues:</h3>
<pre><code>  None</code></pre>
</div>
<div id="included-data-instextdataexample_custom_matrix-and-instextdataexample_fasta_files" class="section level3">
<h3>Included Data (inst\extdata\Example_custom_matrix and inst\extdata\Example_FASTA_files)</h3>
<pre><code>  1.  Two protein sequences in FASTA format (gpr12.fasta, gpr6.fasta)
  2.  A custom matrix developed by Rios et al. (gpcrtm.txt)</code></pre>
</div>
<div id="getting-started" class="section level3">
<h3>Getting started</h3>
<pre><code>(Requires availability of a web browser):

1.  Install R version &gt;= 3.3.0

2.  Install required R packages (and any dependencies):

      'shiny' 
    
      'Biostrings'

3.  Install and load SubVis package

4.  Launch the SubVis application with

      '&gt; startSubVis()'</code></pre>
</div>
<div id="main-tabs-options-and-viz" class="section level3">
<h3>Main tabs: Options and Viz</h3>
<pre><code>After launching, there are two main tabs:  Options and Viz</code></pre>
</div>
<div id="options-tab-loading-data-matrices-and-parameters" class="section level3">
<h3>Options tab: Loading data, matrices, and parameters</h3>
<pre><code>1.  Protein sequences (one per file) must be 
    in FASTA format.  There are two example 
    FASTA files in the test folder.

2.  BLOSUM and PAM matrices can be selected 
    by checking the appropriate box.

3.  A custom matrix in the form of the 
    predefined matrices can be loaded from 
    a text file.  The basic form is 
  
      First row --&gt; Characters representing lookup table.
  
      First col --&gt; Transpose of first row starting at row 2
    
    All other entries are substitution values.
    
    In the display, the custom matrix is labeled &quot;CUST&quot;.
    

4.  The extension penalty, gap penalty, and score 
    type can be selected by text boxes and a drop-down 
    menu.

Note: Each time a change in the options tab is made, 
      the GO button (in the Overview or detail view) 
      must be clicked.</code></pre>
</div>
<div id="viz-tab-overview-and-detail-view" class="section level3">
<h3>Viz tab: Overview and Detail view</h3>
</div>
<div id="overview" class="section level3">
<h3>Overview</h3>
<pre><code>Each row (except the last) represents a percent identity 
type.  Within each row, the pid for the selected matrices 
are sorted.  The last row is the sorted overall score.  

Interaction:  Display initiation
  
  1.  Click 'Go' to initiate display.

Interaction:  Mouse move

  1.  When the mouse moves over a matrix type, 
      the corresponding matrix in the other 
      rows are identified.

Interaction:  Size

  1.  '+' enlarges the display and '-' shrinks the 
      display. (See known issues.)</code></pre>
</div>
<div id="detail" class="section level3">
<h3>Detail</h3>
<pre><code>Score is located under the matrix type for 
each alignment pair.  Amino acids are represented
by a color-coded box.

Interaction:  Size

  1.  '+' enlarges the display.

  2.  '-' shrinks the display.

Interaction:  Mouse over an amino acid.

  1.  Histogram showing the number of each amino 
      acid in that column.

  2.  Log-odds score and the amino acid substitution 
      displayed under the score for each alignment.

  3.  Amino acid, log-odds score, amino acid subsitution 
      is shown in the top right.

Interaction: Classification

  1.  Classify amino acids according to [1] and
      whether an amino acid substitution is conserved (log-odds &gt; 0)
      or not (log-odds &lt; 0)

Interaction: Nagivation

  1.  'Start' - Go to position 1.

  2.  'End' - Go to the end of the maximum alignment.

  3.  'Pair' - Show both the pattern and subject for each matrix type.

  4.  'Patt' - Show only the pattern for each matrix type.

  5.  'Subj' - Show only the subject for each matrix type.

  6.  'Up' - Scroll up if all matrix types will not fit onto display space.

  7.  'Dn' - Scroll down if all matrix types will not fit onto display space.

  8.  '&lt;-' - Go backward in the alignment.

  9.  '-&gt;' - Go forward in the alignment.

  10. 'Alpha' - Show amino acid instead of boxes

  11. 'Pos' - Go to a position by entering column in text box. (Requires clicking 'Go')

Interaction: Search (All require clicking 'Go')

  1.  'NONE' - No search function on (default).

  2.  'INDEL' - Shows the insertions and deletions in red.

  3.  'MATCH' - Shows the positions that match in both the pattern and subject.

  4.  'SEQ' - The amino acid to be searched for.</code></pre>
</div>
<div id="notes" class="section level3">
<h3>Notes:</h3>
<pre><code>1.  Log-odds score for pairs involving a gap are
    reported as undefined.

2.  Most error checking is delegated to the individual 
    R and Biostrings functions.  For example, it is the
    user's responsibility to ensure a valid custom
    matrix.</code></pre>
</div>
<div id="references" class="section level3">
<h3>References:</h3>
<pre><code>1.  Pages, H., Aboyoun, P., Gentleman, R., DebRoy, S.: Biostrings: String
    Objects Representing Biological Sequences, and Matching Algorithms.
    R package version 2.32.0

2.  Pearson, W.R., Lipman, D.J.: Improved tools for biological sequence
    comparison. Proc. Natl. Acad. Sci. USA 85(8), 2444{2448 (1988)

3.  Pommie, C., Levadoux, S., Sabatier, R., Lefranc, G., Lefranc, M.: Imgt
    standardized criteria for statistical analysis of immunoglobulin v-region
    amino acid properties. Journal of Molecular Recognition 17(1), 17{32
    (2004)

4.  R Core Team: R: A Language and Environment for Statistical
    Computing. R Foundation for Statistical Computing, Vienna, Austria
    (2013). R Foundation for Statistical Computing.
    http://www.R-project.org/

5.  RStudio, Inc.: Shiny: Web Application Framework for R. (2014). R
    package version 0.10.1. http://CRAN.R-project.org/package=shiny
    
6.  Rios, S., Fernandez, M.F., Caltabiano, G., Campillo, M., Pardo, L.,
    Gonzalez, A.: Gpcrtm: An amino acid substitution matrix for the
    transmembrane region of class a g protein-coupled receptors. BMC
    Bioinformatics 16(206) (2015)</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
